### PR TITLE
dbuf_hold_impl() cleanup to improve cached read performance

### PR DESCRIPTION
**Pull request by tonynguien**
### Motivation and Context

Improve cached read performance

### Description

Currently every dbuf_hold_impl() incurs kmem_alloc() and kmem_free() which can be costly for cached read performance.

This change reverts the dbuf_hold_impl() fix stack commit, i.e. fc5bb51f08a6c91ff9ad3559d0266eeeab0b1f61 to eliminate the extra kmem_alloc() and kmem_free() operations and improve cached read performance. With the change, each dbuf_hold_impl() frame uses 40 bytes more, total of 800 for 20 recursive levels. Linux kernel stack sizes are 8K and 16K for 32bit and 64bit, respectively, so stack overrun risk is limited.

Sample stack output comparisons with 50 PB file and recordsize=512
Current code
```
 11)     2240      64   arc_alloc_buf+0x4a/0xd0 [zfs]
 12)     2176     264   dbuf_read_impl.constprop.16+0x2e3/0x7f0 [zfs]
 13)     1912     120   dbuf_read+0xe5/0x520 [zfs]
 14)     1792      56   dbuf_hold_impl_arg+0x572/0x630 [zfs]
 15)     1736      64   dbuf_hold_impl_arg+0x508/0x630 [zfs]
 16)     1672      64   dbuf_hold_impl_arg+0x508/0x630 [zfs]
 17)     1608      40   dbuf_hold_impl+0x23/0x40 [zfs]
 18)     1568      40   dbuf_hold_level+0x32/0x60 [zfs]
 19)     1528      16   dbuf_hold+0x16/0x20 [zfs]
```
dbuf_hold_impl() cleanup
```
 11)     2320      64   arc_alloc_buf+0x4a/0xd0 [zfs]
 12)     2256     264   dbuf_read_impl.constprop.17+0x2e3/0x7f0 [zfs]
 13)     1992     120   dbuf_read+0xe5/0x520 [zfs]
 14)     1872      96   dbuf_hold_impl+0x50f/0x5e0 [zfs]
 15)     1776     104   dbuf_hold_impl+0x4df/0x5e0 [zfs]
 16)     1672     104   dbuf_hold_impl+0x4df/0x5e0 [zfs]
 17)     1568      40   dbuf_hold_level+0x32/0x60 [zfs]
 18)     1528      16   dbuf_hold+0x16/0x20 [zfs]
```
### How Has This Been Tested?

Performance observations on 8K recordsize filesystem:
- 8/128/1024K at 1-128 sequential cached read, ~3% improvement

Testing done on Ubuntu 18.04 with 4.15 kernel, 8vCPUs and SSD storage on
VMware ESX.
### Types of changes

- [ ] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [x] Performance enhancement (non-breaking change which improves efficiency)
- [ ] Code cleanup (non-breaking change which makes code smaller or more readable)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)
- [ ] Documentation (a change to man pages or other documentation)

### Checklist:

- [x] My code follows the ZFS on Linux [code style requirements](https://github.com/zfsonlinux/zfs/blob/master/.github/CONTRIBUTING.md#coding-conventions).
- [ ] I have updated the documentation accordingly.
- [x] I have read the [**contributing** document](https://github.com/zfsonlinux/zfs/blob/master/.github/CONTRIBUTING.md).
- [ ] I have added [tests](https://github.com/zfsonlinux/zfs/tree/master/tests) to cover my changes.
- [ ] All new and existing tests passed.
- [x] All commit messages are properly formatted and contain [`Signed-off-by`](https://github.com/zfsonlinux/zfs/blob/master/.github/CONTRIBUTING.md#signed-off-by).

### CODE REVIEW

**Comment by ahrens** [comment]
Looks like there are 2 new test failures which appear to be related to this change:

```
Test: /usr/share/zfs/zfs-tests/tests/functional/arc/dbufstats_001_pos (run as root) [00:00] [FAIL]
00:40:43.31 ASSERTION: dbufstats produces correct statistics
00:40:43.35 SUCCESS: file_write -o create -f /mnt/testdir/file -b 1048576 -c 20 -d R
00:40:43.49 SUCCESS: zpool sync
00:40:43.76 SUCCESS: eval cat /proc/spl/kstat/zfs/dbufs > /mnt/dbufs.out.DD4gCe
00:40:43.76 SUCCESS: eval cat /proc/spl/kstat/zfs/dbufstats > /mnt/dbufstats.out.37w2eH
00:40:43.83 NOTE: Checking if 4126 is within +/-15 of 193 (diff: 3933)
00:40:43.83 Stat cache_level_0 exceeded tolerance
```
```
Test: /usr/share/zfs/zfs-tests/tests/functional/arc/dbufstats_002_pos (run as root) [00:00] [FAIL]
00:40:43.88 ASSERTION: dbufs move from mru to mfu list
00:40:43.89 SUCCESS: file_write -o create -f /mnt/testdir/file -b 1048576 -c 1 -d R
00:40:43.92 SUCCESS: zpool sync
00:40:43.92 NOTE: Object ID for /mnt/testdir/file is 3
00:40:44.19 SUCCESS: eval cat /proc/spl/kstat/zfs/dbufs > /mnt/dbufs.out.aYZTIA
00:40:44.33 NOTE: dbuf count is 9, mru count is 9, mfu count is 0
00:40:44.33 SUCCESS: eval cat /mnt/testdir/file > /dev/null
00:40:44.59 SUCCESS: eval cat /proc/spl/kstat/zfs/dbufs > /mnt/dbufs.out.aYZTIA
00:40:44.73 NOTE: dbuf count is 9, mru count is 9, mfu count is 0
00:40:44.73 Compared mfu count should be not equal: 0 == 0
```

**Comment by tonynguien** [comment]
Hmm. I'll look into those failures. Thanks!

File: module/zfs/dbuf.c [hunk 1]
```diff
@@ -3269,178 +3241,111 @@ dbuf_hold_copy(struct dbuf_hold_arg *dh)
 	}
 
 	rw_enter(&db->db_rwlock, RW_WRITER);
-	bcopy(data->b_data, db->db.db_data, arc_buf_size(data));
+	bcopy(dr->dt.dl.dr_data->b_data, db->db.db_data, db->db.db_size);
```
Comment (behlendorf) [inline module/zfs/dbuf.c]: I believe you do want to use `arc_buf_size(data)` here.

File: module/zfs/dbuf.c [hunk 2]
```diff
@@ -3269,178 +3241,111 @@ dbuf_hold_copy(struct dbuf_hold_arg *dh)
...
 
-	if (dh->dh_fail_uncached && dh->dh_db->db_state != DB_CACHED) {
-		mutex_exit(&dh->dh_db->db_mtx);
+	if (fail_uncached && db->db_state != DB_CACHED) {
+		mutex_exit(&db->db_mtx);
 		return (SET_ERROR(ENOENT));
 	}
 
-	if (dh->dh_db->db_buf != NULL) {
-		arc_buf_access(dh->dh_db->db_buf);
```
Comment (behlendorf) [inline module/zfs/dbuf.c]: You need to add back this call to `arc_buf_access` which was dropped.  It's very likely responsible for the failed test cases, see 0873bb6337452e3e028e40f5dad945b30deab185.

File: module/zfs/dbuf.c [hunk 3]
```diff
@@ -3269,178 +3241,111 @@ dbuf_hold_copy(struct dbuf_hold_arg *dh)
...
+		    db->db.db_size, db);
 
-		if (dh->dh_db->db_caching_status == DB_DBUF_METADATA_CACHE) {
-			DBUF_STAT_BUMPDOWN(metadata_cache_count);
-		} else {
-			DBUF_STAT_BUMPDOWN(cache_levels[dh->dh_db->db_level]);
-			DBUF_STAT_BUMPDOWN(cache_count);
-			DBUF_STAT_DECR(cache_levels_bytes[dh->dh_db->db_level],
-			    dh->dh_db->db.db_size);
-		}
```
Comment (behlendorf) [inline module/zfs/dbuf.c]: These stats should be preserved.

File: module/zfs/dbuf.c (hunk 2 above)
Comment (tonynguien) [reply module/zfs/dbuf.c]: Right. Will add that back.

File: module/zfs/dbuf.c (hunk 3 above)
Comment (tonynguien) [reply module/zfs/dbuf.c]: Will do.

File: module/zfs/dbuf.c (hunk 1 above)
Comment (tonynguien) [reply module/zfs/dbuf.c]: Right, I will use `arc_buf_size(data)`.

**Comment by tonynguien** [comment]
I made the suggested changes and reran the tests. The performance gain smaller, ~3%, with `arc_buf_access()` in place. I've revised the commit message and PR description.

File: module/zfs/dbuf.c [hunk 4]
```diff
@@ -3277,170 +3249,113 @@ dbuf_hold_copy(struct dbuf_hold_arg *dh)
...
+		arc_buf_access(db->db_buf);
+		ASSERT3P(db->db.db_data, ==, db->db_buf->b_data);
 	}
 
-	ASSERT(dh->dh_db->db_buf == NULL || arc_referenced(dh->dh_db->db_buf));
+	ASSERT(db->db_buf == NULL || arc_referenced(db->db_buf));
 
 	/*
-	 * If this buffer is currently syncing out, and we are
+	 * If this buffer is currently syncing out, and we are are
```
Comment (behlendorf) [inline module/zfs/dbuf.c]: nit: are are

Comment (tonynguien) [reply module/zfs/dbuf.c]: Tiny nit corrected. Thanks Brian!

**Pull request merged**